src/mlpack/tests/main_tests/kfn_test.cpp

```diff
@@ -0,0 +1,378 @@
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("tree_type", (string) "min-rp"); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/*
+ * Check that we can't pass an invalid algorithm.
+ */
+BOOST_AUTO_TEST_CASE(KFNInvalidAlgoTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("algorithm", (string) "triple_tree"); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/*
+ * Check that we can't pass an invalid value of epsilon.
+ */
+BOOST_AUTO_TEST_CASE(KFNInvalidEpsilonTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("epsilon", (double) -1); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Make sure that dimensions of the neighbors and distances
+ * matrices are correct given a value of k.  
+ */
+BOOST_AUTO_TEST_CASE(KFNOutputDimensionTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+  
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  
+  mlpackMain();
+
+  // Check the neighbors matrix has 4 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>("neighbors").n_rows, 10);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>("neighbors").n_cols, 100);
+
+  // Check the distances matrix has 4 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_rows, 10);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_cols, 100);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(KFNModelReuseTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  arma::mat queryData;
+  queryData.randu(3, 90); // 90 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("query", queryData);
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighbors = std::move(CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  distances = std::move(CLI::GetParam<arma::mat>("distances"));
+  
```

Comment:
I think we need to call `CleanMemory()` before a second call of `mlpackMain()`, otherwise the program may have allocated a `KFNModel` that needs to be deleted.

Reply:
calling CleanMemory() throws a `memory access violation`. I am currently trying to dig into this and the other segmentation fault with a debugger.

Reply:
This is likely because of the line down below:

```
SetInputParam("input_model", 
      std::move(CLI::GetParam<KFNModel*>("output_model")));
```

Before calling `CleanMemory()` you'll need to copy the `output_model` result, since `CleanMemory()` will delete it.

Reply:
I had a related doubt, after calling `CleanMemory()`, if we don't call `mlpackMain()`, why is a `memory access violation` thrown?
Also, I noticed that otherwise this error is only thrown whenever we set the input parameter `input_model` both before or after the CleanMemory() call.